libdevcrypto/LibSnark.cpp

```diff
@@ -26,54 +23,51 @@
 #include <algebra/curves/alt_bn128/alt_bn128_pp.hpp>
 #include <common/profiling.hpp>
 
-#include <libdevcrypto/Exceptions.h>
-
-#include <libdevcore/CommonIO.h>
+#include <libdevcore/Exceptions.h>
 #include <libdevcore/Log.h>
 
-#include <mutex>
-
 using namespace std;
 using namespace dev;
 using namespace dev::crypto;
 
 namespace
 {
 
-std::once_flag initLibSnarkFlag;
+DEV_SIMPLE_EXCEPTION(InvalidEncoding);
 
-void initLibSnark()
+void initLibSnark() noexcept
 {
-	call_once(initLibSnarkFlag,
-		[](){
-			cnote << "initializing libsnark.";
-			libff::inhibit_profiling_info = true;
-			libff::inhibit_profiling_counters = true;
-			libff::alt_bn128_pp::init_public_params();
-		});
+	static bool s_initialized = []() noexcept
+	{
+		libff::inhibit_profiling_info = true;
+		libff::inhibit_profiling_counters = true;
+		libff::alt_bn128_pp::init_public_params();
+		return true;
+	}();
+	(void)s_initialized;
```

Comment:
We need any `static` variable to use thread-safe static initialization (aka magic statics). We can use `call_once` here explicitly, but static vars are handled better by compilers.
What do you think? Go back to `call_once`?